File: java/src/main/java/mchehab/com/java/MainActivity.java
package mchehab.com.java;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.MarkerOptions;
import com.google.maps.android.clustering.ClusterManager;
import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class MainActivity extends AppCompatActivity implements OnMapReadyCallback {

    private GoogleMap googleMap;
    private JSONArray jsonArray;
    private List<MarkerOptions> listMarkers = new ArrayList<>();
    private ClusterManager<MarkerClusterItem> clusterManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        jsonArray = readAssets();

        SupportMapFragment supportMapFragment = (SupportMapFragment)getSupportFragmentManager().findFragmentById(R.id.googleMap);
        supportMapFragment.getMapAsync(this);
    }

    private JSONArray readAssets(){
        try{
            BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(getAssets().open("places.json")));
            String line = "";
            StringBuilder stringBuilder = new StringBuilder("");
            while ((line = bufferedReader.readLine()) != null){
                stringBuilder.append(line);
            }
            return new JSONArray(stringBuilder.toString());
        }catch (Exception e){
            e.printStackTrace();
        }
        return new JSONArray();
    }

    private void addMarkers(){
        try{
            for (int i = 0; i < jsonArray.length(); i++){
                JSONObject jsonObject = jsonArray.getJSONObject(i);
                double lat = jsonObject.getDouble("lat");
                double lng = jsonObject.getDouble("lng");
                String name = jsonObject.getString("name");
                LatLng latLng = new LatLng(lat, lng);
                MarkerOptions markerOptions = new MarkerOptions().position(latLng).title(name);
                listMarkers.add(markerOptions);
            }
            googleMap.animateCamera(CameraUpdateFactory.newLatLngZoom(listMarkers.get(0).getPosition(), 13.0f));
        }catch (Exception e){
            e.printStackTrace();
        }
    }

    private void setupClusterManager() {
        setRenderer();
        addClusterItems();
        setClusterManagerClickListener();
        googleMap.setOnCameraIdleListener(clusterManager);
        googleMap.setOnMarkerClickListener(clusterManager);
        clusterManager.cluster();
    }

    private void setRenderer() {
        MarkerClusterRenderer<MarkerClusterItem> clusterRenderer = new MarkerClusterRenderer<>(this, googleMap, clusterManager);
        clusterManager.setRenderer(clusterRenderer);
    }

    private void setClusterManagerClickListener() {
        clusterManager.setOnClusterClickListener(cluster -> {
            Collection<MarkerClusterItem> listItems = cluster.getItems();
            List<String> listNames = new ArrayList<>();
            for (MarkerClusterItem item : listItems){
                listNames.add(item.getTitle());
            }
            googleMap.animateCamera(CameraUpdateFactory.newLatLng(cluster.getPosition()), new GoogleMap.CancelableCallback() {
                @Override
                public void onFinish() {
                    ListViewDialog listViewDialog = new ListViewDialog(MainActivity.this, listNames);
                    listViewDialog.showDialog();
                }
                @Override
                public void onCancel() { }
            });
            return true;
        });
    }

    private void addClusterItems() {
        for(MarkerOptions markerOptions : listMarkers){
            MarkerClusterItem clusterItem = new MarkerClusterItem(markerOptions.getPosition(), markerOptions.getTitle());
            clusterManager.addItem(clusterItem);
        }
    }

    @Override
    public void onMapReady(GoogleMap googleMap) {
        this.googleMap = googleMap;
        clusterManager = new ClusterManager<>(this, googleMap);
        addMarkers();
        setupClusterManager();
    }
}

File: app/src/main/java/mchehab/com/googlemapsmarkergrouping/MainActivity.kt
package mchehab.com.googlemapsmarkergrouping

import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MarkerOptions
import com.google.maps.android.clustering.ClusterItem
import com.google.maps.android.clustering.ClusterManager
import org.json.JSONArray


class MainActivity : AppCompatActivity(), OnMapReadyCallback{

    private lateinit var googleMap: GoogleMap
    private lateinit var jsonArray: JSONArray
    private lateinit var clusterManager: ClusterManager<ClusterItem>
    private val markerList = mutableListOf<MarkerOptions>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        jsonArray = readAssets()

        val supportFragment = supportFragmentManager.findFragmentById(R.id.googleMap) as SupportMapFragment
        supportFragment.getMapAsync(this)
    }

    private fun readAssets(): JSONArray{
        val json = assets.open("places.json")
            .bufferedReader()
            .readText()
        return JSONArray(json)
    }

    private fun addMarkers(){
        for (index in 0 until jsonArray.length()){
            val jsonObject = jsonArray.getJSONObject(index)
            val name = jsonObject.getString("name")
            val lat = jsonObject.getDouble("lat")
            val lng = jsonObject.getDouble("lng")
            val marker = MarkerOptions().position(LatLng(lat, lng)).title(name)
            markerList.add(marker)
        }
        googleMap.animateCamera(CameraUpdateFactory.newLatLngZoom(markerList[0].position, 13.0f))
    }

    override fun onMapReady(googleMap: GoogleMap) {
        this.googleMap = googleMap
        clusterManager = ClusterManager(this, googleMap)
        addMarkers()
        setupClusterManager()
    }

    private fun setupClusterManager() {
        clusterManager.renderer = MarkerClusterRenderer(this, googleMap, clusterManager)
        addClusters()
        setClusterManagerClickListener()
        googleMap.setOnCameraIdleListener(clusterManager)
        googleMap.setOnMarkerClickListener(clusterManager)
        clusterManager.cluster()
    }

    private fun addClusters() {
        for (marker in markerList){
            val clusterItem = MarkerClusterItem(marker.position, marker.title)
            clusterManager.addItem(clusterItem)
        }
    }

    private fun setClusterManagerClickListener() {
        clusterManager.setOnClusterClickListener {
            val items = it.items
            val itemsList = mutableListOf<String>()
            for (item in items){
                itemsList.add((item as MarkerClusterItem).markerTitle)
            }
            googleMap.animateCamera(CameraUpdateFactory.newLatLng(it.position), object: GoogleMap.CancelableCallback{
                override fun onFinish() {
                    ListViewDialog(this@MainActivity, itemsList)
                }
                override fun onCancel() {}
            })
            true
        }
    }
}